### PR TITLE
chore(awslint): new rules for reference interfaces

### PR DESCRIPTION
**Pull request by mrgrain**
Introducing two new rules, and immediately accepting all current violations.

- `resource-interface-extends-resource-ref` - all resources interfaces like `IBucket` must also implement the reference interface, e.g. `ICfnBucket` 
- `ref-via-ref-interface` - all APIs should accept the reference interface `ICfnBucket` over the resource interface `IBucket` 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/awslint/lib/rules/api.ts [hunk 1]
```diff
@@ -3,143 +3,220 @@ import { ConstructReflection } from './construct';
...
+});
 
 apiLinter.add({
   code: 'ref-via-interface',
-  message: 'API should use interface and not the concrete class (%s). ' +
+  message:
+    'API should use interface and not the concrete class (%s). ' +
     `If this is intentional, add "${EXCLUDE_ANNOTATION_REF_VIA_INTERFACE}" to element's jsdoc`,
-  eval: e => {
+  eval: (e) => {
```
Comment (mrgrain) [inline packages/awslint/lib/rules/api.ts:53]: This rule hasn't changed. I just refactored the code used to visit spec nodes to be reusable.

File: packages/awslint/lib/rules/api.ts [hunk 2]
```diff
@@ -3,143 +3,220 @@ import { ConstructReflection } from './construct';
 import { CoreTypes } from './core-types';
 import { Linter } from '../linter';
 
-const EXCLUDE_ANNOTATION_REF_VIA_INTERFACE = '[disable-awslint:ref-via-interface]';
+const EXCLUDE_ANNOTATION_REF_VIA_INTERFACE =
+  '[disable-awslint:ref-via-interface]';
+
+const EXCLUDE_ANNOTATION_REF_VIA_REF_INTERFACE =
+  '[disable-awslint:ref-via-ref-interface]';
```
Comment (mrgrain) [inline packages/awslint/lib/rules/api.ts:10]: I considered wrapping them both into a single rule, but found it too confusing and complicated.
Now we have two rules:
- `ref-via-interface` guarantees that we are not expecting concrete instances of constructs
- `ref-via-ref-interface` guarantees that we prefer `ICfnWhatever` reference interfaces over `IWhatever` interfaces

**Comment by mikewrighton** [comment]
Generally looks good to me - did you see what impact this had on the lint times?

File: packages/awslint/lib/rules/construct.ts [hunk 3]
```diff
@@ -184,6 +184,8 @@ constructLinter.add({
   message: 'construct interface must extend core.IConstruct',
   eval: e => {
     if (!e.ctx.interfaceType) { return; }
+    // ICfn interfaces are not construct interfaces
```
Comment (rix0rrr) [inline packages/awslint/lib/rules/construct.ts:187]: I wonder if they should be though? `IBucket extends IConstruct`, I wonder if we should `ICfnBucket extends IConstruct` as well.

Comment (mrgrain) [reply packages/awslint/lib/rules/construct.ts:187]: @kaizencc

**Review by mikewrighton — APPROVED** [review]

**Comment by mrgrain** [comment]
Adds about 5s to awslint. Deeming this as acceptable. I am working on some speed-ups for the linter.

**Pull request merged**